src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java

```diff
@@ -322,10 +322,58 @@ public void remove(Long id) throws ValidationException, UserNotFoundException, N
         rawUserRepository.delete(user.getId());
     }
 
+    @Override
+    public List<U> createAll(final @NotNull List<U> users, boolean emailConfirmationRequired, String registrantToken, String callbackUrl)
+            throws PasswordValidationException {
+
+        for (U user : users) {
+            updateFields(user);
+        }
+
+        List<U> createdUsers = userRepository.save(users);
+        if (emailConfirmationRequired) {
+            Optional<UserRegistrant> userRegistrant = userRegistrantService.findByToken(registrantToken);
+            for (U user : createdUsers) {
+                sendRegistrationEmail(user, userRegistrant, callbackUrl, true);
+            }
+        }
+
+        return createdUsers;
+    }
+
     @Override
     public U create(final @NotNull U user)
             throws NotUniqueException, NotExistException, PasswordValidationException {
 
+        updateFields(user);
+
+        return userRepository.save(user);
+    }
+
+    @Override
+    public U createWithEmailVerification(final @NotNull U user, String registrantToken, String callbackUrl)
+            throws NotUniqueException, NotExistException, PasswordValidationException {
+
+        user.setEmailConfirmed(false);
+        user.setRegistrationCode(UUID.randomUUID().toString());
+        U createdUser = create(user);
+
+        Optional<UserRegistrant> userRegistrant = userRegistrantService.findByToken(registrantToken);
+        sendRegistrationEmail(createdUser, userRegistrant, callbackUrl, false);
+        return createdUser;
+    }
+
+    private void updateFields(U user) throws PasswordValidationException {
+        // The existing user check should come last:
+        // it is muted in public registration form, so we need to show other errors ahead
+        U byEmail = getByUnverifiedEmailInAnyTenant(user.getEmail());
```

Comment:
Why the code block which should come last (and this is described in the comment) was moved to the start of the method?